### DISCUSSION TITLE
how to debug undefined symbol

### DISCUSSION THREAD

**Discussion by jwuttke** [RosettaCommons/binder#188] [Q&A] [answered]
Say, definition of class `A` in A.h contains a function `f17` that is not implemented in A.cpp.

Binder will pass, pybind11 will pass, compiler will pass, linker will pass - but the Python import statement will fail with a message like
```
ImportError: my_module.cpython-39-x86_64-linux-gnu.so: undefined symbol: _ZNK1A2a2Ev
```
How to debug this error? Is there a way to make the error message more explicit? Is there a way to catch this error earlier?

**Comment by lyskov** [comment, marked as answer]
This is indeed an annoying issue. Simplest way to debug this would to use UNIX `nm` utility. I usually do something like `nm -D --demangle --undefined-only <your-project-lib>.so | grep <primary-namespace>`. Also (and sometimes this a bit easier) linking on MacOS will immediately fail if there is unresolved symbols.
